src/main/java/org/apache/commons/lang3/reflect/TypeUtils.java

```diff
@@ -1823,13 +1840,16 @@ private static void appendRecursiveTypes(final StringBuilder buf, final int[] re
         }
     }
 
-    private static int[] findRecursiveTypes(final ParameterizedType p) {
+    @SuppressWarnings("assignment.type.incompatible")/*
+    indexesToRemove = ArrayUtils.add(indexesToRemove, i); // ArrayUtils.add(indexesToRemove, i) at the end of the loop returns an array with maximum possible length of p.getActualTypeArguments().length 
+    */
+    private static int @LTEqLengthOf("#1.getActualTypeArguments()") [] findRecursiveTypes(final ParameterizedType p) {
         final Type[] filteredArgumentTypes = Arrays.copyOf(p.getActualTypeArguments(), p.getActualTypeArguments().length);
-        int[] indexesToRemove = {};
+        int @LTEqLengthOf("p.getActualTypeArguments()") [] indexesToRemove = {};
         for (int i = 0; i < filteredArgumentTypes.length; i++) {
             if (filteredArgumentTypes[i] instanceof TypeVariable<?>) {
                 if (containsVariableTypeSameParametrizedTypeBound(((TypeVariable<?>) filteredArgumentTypes[i]), p)) {
-                    indexesToRemove = ArrayUtils.add(indexesToRemove, i);
+                    indexesToRemove = ArrayUtils.add(indexesToRemove, i); // ArrayUtils.add(indexesToRemove, i) at the end of the loop returns an array with maximum possible length of p.getActualTypeArguments().length 
```

Comment:
How do we know that `i` is an index for `p.getActualTypeArguments()`? That seems to be the reason that this error is being issued, but your justification doesn't mention that.

Reply:
Related, is there an annotation we could write that would express that fact?